### PR TITLE
init: create default spidermultusconfig network if it isn't empty

### PR DESCRIPTION
**Pull request by cyclinder**
---
name: Pull request
about: Tell us about your contribution
labels: ["pr/release/none-required"]
---

---
## Thanks for contributing !

Before submitting a pull request, make sure you read about our Contribution notice here: <https://spidernet-io.github.io/spiderpool/latest/develop/contributing/>

**What this PR does / why we need it**:

If `multus.defaultCNICrName` is set when spiderpool installing, We should create default SpidermultusConfig instance for it

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**make sure your commit is signed off**

### CODE REVIEW

File: charts/spiderpool/templates/multus/multus-daemonset.yaml [hunk 1]
```diff
@@ -46,12 +46,11 @@ kind: ServiceAccount
 metadata:
   name: {{ .Values.multus.multusCNI.name | trunc 63 | trimSuffix "-" }}
   namespace: {{ .Release.Namespace | quote }}
-{{- if .Values.multus.multusCNI.defaultCniCRName }}
 ---
 kind: ConfigMap
 apiVersion: v1
 metadata:
-  name: multus-cni-config
```
Comment (weizhoublue) [inline charts/spiderpool/templates/multus/multus-daemonset.yaml:54]: 建议不需要这么多定制，如果要换名字，可以复用
{{ .Values.multus.multusCNI.name | trunc 63 | trimSuffix "-" }}

File: cmd/spiderpool-init/cmd/root.go [hunk 2]
```diff
@@ -5,21 +5,48 @@ package cmd
...
+	"github.com/spidernet-io/spiderpool/pkg/multuscniconfig"
 	"github.com/spidernet-io/spiderpool/pkg/utils"
 )
 
 var logger *zap.Logger
 
+// MultusNetConf for cni config file written in json
+// Note: please keep this fields be consistent with multus configMap
+// in charts/spiderpool/templates/multus/multus-daemonset.yaml
+type MultusNetConf struct {
```
Comment (weizhoublue) [inline cmd/spiderpool-init/cmd/root.go:31]: 这个是否可直接 引用 multus 的 pkg/config/generator.go

File: cmd/spiderpool-init/cmd/root.go [hunk 3]
```diff
@@ -192,5 +213,62 @@ func InitDefaultMultusCNIConfig(ctx context.Context, client *CoreClient, cniDir
...
+	configDatas, err := json.Marshal(multusConfig)
+	if err != nil {
+		return fmt.Errorf("failed to marshal multus config: %v", err)
+	}
+	cm.Data["cni-conf.json"] = string(configDatas)
+
+	logger.Sugar().Infof("Patch multus configMap %s: %s", multusConfigMap, configDatas)
+
+	// patch
+	if err = client.Patch(ctx, cm, controller_client.MergeFrom(oldConfigMap)); err != nil {
```
Comment (weizhoublue) [inline cmd/spiderpool-init/cmd/root.go:250]: patch 完后，需要重启 multus daemonset ？

File: charts/spiderpool/templates/multus/multus-daemonset.yaml [hunk 4]
```diff
@@ -46,12 +46,11 @@ kind: ServiceAccount
 metadata:
   name: {{ .Values.multus.multusCNI.name | trunc 63 | trimSuffix "-" }}
```
Comment (weizhoublue) [inline charts/spiderpool/templates/multus/multus-daemonset.yaml:47]: 我们之前谈到，是否可以把 它的这个 daemonset 合入为 spiderpool 的 一个 init pod ？

File: cmd/spiderpool-init/cmd/root.go [hunk 5]
```diff
@@ -151,8 +178,13 @@ func Execute() {
 
 	// create multuscniconfig for default network
 	if config.DefaultCNIName == "" {
-		logger.Sugar().Infof("Try to create MultusCniConfig default network in %s", config.DefaultCNIDir)
-		if err = InitDefaultMultusCNIConfig(ctx, client, config.DefaultCNIDir, config.DefaultCNINamespace); err != nil {
+		logger.Sugar().Infof("multus.defaultCniCRName is empty, try to create Spidermultusconfig default network in %s", config.DefaultCNIDir)
```
Comment (weizhoublue) [inline cmd/spiderpool-init/cmd/root.go:181]: config.DefaultCNIName 是否为空，形成了 2 套逻辑
其实是否可以复用的，保障一些行为是一致的：
（1）config.DefaultCNIName 为空也好，不为空也好，计算返回一个 cr name
（2）确认该 cr 是否被创建
（3）确认 cm 中是该 name，需要则重启它

File: cmd/spiderpool-init/cmd/root.go (hunk 2 above)
Comment (cyclinder) [reply cmd/spiderpool-init/cmd/root.go:31]: 试了不行，有依赖冲突

File: charts/spiderpool/templates/multus/multus-daemonset.yaml (hunk 4 above)
Comment (cyclinder) [reply charts/spiderpool/templates/multus/multus-daemonset.yaml:47]: 这个在下一个 PR 统一做

File: cmd/spiderpool-init/cmd/root.go [hunk 6]
```diff
@@ -192,5 +213,62 @@ func InitDefaultMultusCNIConfig(ctx context.Context, client *CoreClient, cniDir
...
 
+	// patch configMap
+	if err = patchMultusConfigMap(ctx, client, multusConfigMap, cniName, ns); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func patchMultusConfigMap(ctx context.Context, client *CoreClient, multusConfigMap, defaultNetwork, ns string) error {
```
[the remark this answers is not in the document]
Comment (cyclinder) [reply cmd/spiderpool-init/cmd/root.go:224]: 可

File: cmd/spiderpool-init/cmd/root.go (hunk 3 above)
Comment (cyclinder) [reply cmd/spiderpool-init/cmd/root.go:250]: 是需要的

File: charts/spiderpool/templates/multus/multus-daemonset.yaml (hunk 4 above)
Comment (cyclinder) [reply charts/spiderpool/templates/multus/multus-daemonset.yaml]: Done

File: cmd/spiderpool-init/cmd/multus.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	var err error
+	if err = client.Get(ctx, apitypes.NamespacedName{Name: name, Namespace: ns}, &spiderAgent); err != nil {
+		return err
+	}
+
+	if err = client.DeleteAllOf(ctx, &corev1.Pod{}, controller_client.InNamespace(ns), controller_client.MatchingLabels(spiderAgent.Spec.Template.Labels)); err != nil {
+		return err
+	}
+
+	if err = client.WaitPodListReady(ctx, ns, spiderAgent.Spec.Template.Labels); err != nil {
```
Comment (weizhoublue) [inline cmd/spiderpool-init/cmd/multus.go:111]: 这里可能有 bug， 马上 delete， 马上就 wait， 可能pod还来得及真正 kill，已经 判定 ready 了
需要有个 sleep

File: cmd/spiderpool-init/cmd/multus.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+func restartSpiderAgent(ctx context.Context, client *CoreClient, name, ns string) error {
+	logger.Sugar().Infof("Try to restart spiderpoo-agent daemonSet: %s/%s", ns, name)
+
+	var spiderAgent v1.DaemonSet
+	var err error
+	if err = client.Get(ctx, apitypes.NamespacedName{Name: name, Namespace: ns}, &spiderAgent); err != nil {
+		return err
+	}
+
+	if err = client.DeleteAllOf(ctx, &corev1.Pod{}, controller_client.InNamespace(ns), controller_client.MatchingLabels(spiderAgent.Spec.Template.Labels)); err != nil {
```
Comment (weizhoublue) [inline cmd/spiderpool-init/cmd/multus.go:107]: delete 好 还是 滚动更新 deaemonset 之类的 api 好

File: cmd/spiderpool-init/cmd/root.go [hunk 9]
```diff
@@ -149,48 +146,27 @@ func Execute() {
...
-		return fmt.Errorf("failed to parseCNIFromConfig: %v", err)
+	// patch configMap
+	if err = patchMultusConfigMap(ctx, client, config.MultusConfigMap, defaultCNIName, config.DefaultCNINamespace); err != nil {
+		logger.Fatal(err.Error())
 	}
 
-	if err = client.WaitMultusCNIConfigCreated(ctx, getMultusCniConfig(cniName, cniType, ns)); err != nil {
-		return fmt.Errorf("failed to WaitMultusCNIConfigCreated: %v", err)
+	// restart multus pod
+	if err = restartSpiderAgent(ctx, client, config.AgentName, config.DefaultCNINamespace); err != nil {
```
Comment (weizhoublue) [inline cmd/spiderpool-init/cmd/root.go:164]: 这里可以优化下，如果之前 的 cni 或者 configmap  已经是 预期的 default，这里就不需要 重启了，可加速部署

File: cmd/spiderpool-init/cmd/multus.go (hunk 7 above)
Comment (cyclinder) [reply cmd/spiderpool-init/cmd/multus.go:126]: time.Sleep( 10 * time.Second)

File: cmd/spiderpool-init/cmd/root.go (hunk 9 above)
Comment (cyclinder) [reply cmd/spiderpool-init/cmd/root.go]: Done

File: cmd/spiderpool-init/cmd/multus.go (hunk 7 above)
Comment (weizhoublue) [reply cmd/spiderpool-init/cmd/multus.go:126]: 10 是否长了点，3s 即可 ？

**Comment by weizhoublue** [comment]
很多文档中，get pod 后出现的 spiderpool-multus ，一起删除了

File: cmd/spiderpool-init/cmd/multus.go (hunk 7 above)
Comment (weizhoublue) [reply cmd/spiderpool-init/cmd/multus.go:126]: 10 是否长了点，3s 即可 ？

File: cmd/spiderpool-init/cmd/client.go [hunk 10]
```diff
@@ -180,3 +180,39 @@ func (c *CoreClient) WaitMultusCNIConfigCreated(ctx context.Context, multuscnico
...
+				noReady = true
+				break
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			interval := retryIntervalSec * time.Second
```
Comment (weizhoublue) [inline cmd/spiderpool-init/cmd/client.go:212]: 这个 应该要一个超时机制 ， 万一超时了，错误退出

File: cmd/spiderpool-init/cmd/multus.go (hunk 7 above)
Comment (cyclinder) [reply cmd/spiderpool-init/cmd/multus.go:126]: done

**Comment by weizhoublue** [comment]
ci failed

**Comment by cyclinder** [comment]
```
• [FAILED] [300.432 seconds]
test kubevirt [kubevirt]
/home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:26
  [It] Succeed to keep static IP for kubevirt VM/VMI after restarting the VM/VMI pod [F00001]
  /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:59

  Timeline >>
  > Enter [BeforeEach] test kubevirt - /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:33 @ 10/26/23 14:55:46.488
  create namespace nslz5jw5. 
  < Exit [BeforeEach] test kubevirt - /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:33 @ 10/26/23 14:55:46.541 (53ms)
  > Enter [It] Succeed to keep static IP for kubevirt VM/VMI after restarting the VM/VMI pod - /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:59 @ 10/26/23 14:55:46.542
  try to create kubevirt VM: &{{VirtualMachine kubevirt.io/v1} {testvm-r57hr4  nslz5jw5    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[kubevirt.io/vm:vm-cirros] map[] [] [] []} {<nil> 0xc000474070 <nil> <nil> 0xc000740a80 [] <nil>} {<nil> <nil> false false  [] [] [] [] <nil> <nil> 0 0 []}} 
  Automatically polling progress:
    test kubevirt Succeed to keep static IP for kubevirt VM/VMI after restarting the VM/VMI pod (Spec Runtime: 2m0.054s)
      /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:59
      In [It] (Node Runtime: 2m0s)
        /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:59
```

**Comment by cyclinder** [comment]
```
test ifacer [ifacer]
/home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:20
  [It] About ifacer's e2e use cases [N00001, N00002, N00003, N00006]
  /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:101

  Timeline >>
  > Enter [BeforeEach] test ifacer - /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:28 @ 10/26/23 14:37:48.584
  Generate vlan ID of sub-interface: 58
  create namespace ns3748-584404833. 
  Generate spiderMultusConfig &{{ } {test-multus-f3a984c7fd  ns3748-584404833    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []} {macvlan 0xc0007cacf0 <nil> <nil> <nil> <nil> <nil> 0xc000668f00 <nil>}} 
  < Exit [BeforeEach] test ifacer - /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:28 @ 10/26/23 14:37:48.96 (376ms)
  > Enter [It] About ifacer's e2e use cases - /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:101 @ 10/26/23 14:37:48.965
  Create spidermultus config ns3748-584404833/test-multus-f3a984c7fd 
  Try to create Deployment: ns3748-584404833/ds-5e08ac89f3 
  Check that each node where the Pod is located should have a vlan sub-interface.
  docker exec -i spiderpool1026142530-worker ip link show eth0 | grep 'state UP mode'  
  docker exec -i spiderpool1026142530-worker ip link show | grep eth0.58 | grep 'state UP mode'  
  Failed to execute command on the node, error is  
  docker exec -i spiderpool1026142530-worker ip link show eth0 | grep 'state UP mode'  
  docker exec -i spiderpool1026142530-worker ip link show | grep eth0.58 | grep 'state UP mode'  
  docker exec -i spiderpool1026142530-control-plane ip link show eth0 | grep 'state UP mode'  
  docker exec -i spiderpool1026142530-control-plane ip link show | grep eth0.58 | grep 'state UP mode'  
  Create a vlan sub-interface with the same name, its network card status is down, and it is automatically set to up
  docker exec -i spiderpool1026142530-worker ip link set eth0.58 down 
  docker exec -i spiderpool1026142530-control-plane ip link set eth0.58 down 
  [FAILED] Failed to execute down nic command on the node, error is Cannot find device "eth0.58"

```

**Comment by ty-dc** [comment]
> ```
> • [FAILED] [300.432 seconds]
> test kubevirt [kubevirt]
> /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:26
>   [It] Succeed to keep static IP for kubevirt VM/VMI after restarting the VM/VMI pod [F00001]
>   /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:59
> 
>   Timeline >>
>   > Enter [BeforeEach] test kubevirt - /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:33 @ 10/26/23 14:55:46.488
>   create namespace nslz5jw5. 
>   < Exit [BeforeEach] test kubevirt - /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:33 @ 10/26/23 14:55:46.541 (53ms)
>   > Enter [It] Succeed to keep static IP for kubevirt VM/VMI after restarting the VM/VMI pod - /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:59 @ 10/26/23 14:55:46.542
>   try to create kubevirt VM: &{{VirtualMachine kubevirt.io/v1} {testvm-r57hr4  nslz5jw5    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[kubevirt.io/vm:vm-cirros] map[] [] [] []} {<nil> 0xc000474070 <nil> <nil> 0xc000740a80 [] <nil>} {<nil> <nil> false false  [] [] [] [] <nil> <nil> 0 0 []}} 
>   Automatically polling progress:
>     test kubevirt Succeed to keep static IP for kubevirt VM/VMI after restarting the VM/VMI pod (Spec Runtime: 2m0.054s)
>       /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:59
>       In [It] (Node Runtime: 2m0s)
>         /home/runner/work/spiderpool/spiderpool/test/e2e/kubevirt/kubevirt_test.go:59
> ```

Known issues

https://github.com/spidernet-io/spiderpool/pull/2461

**Comment by ty-dc** [comment]
> ```
> test ifacer [ifacer]
> /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:20
>   [It] About ifacer's e2e use cases [N00001, N00002, N00003, N00006]
>   /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:101
> 
>   Timeline >>
>   > Enter [BeforeEach] test ifacer - /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:28 @ 10/26/23 14:37:48.584
>   Generate vlan ID of sub-interface: 58
>   create namespace ns3748-584404833. 
>   Generate spiderMultusConfig &{{ } {test-multus-f3a984c7fd  ns3748-584404833    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] [] []} {macvlan 0xc0007cacf0 <nil> <nil> <nil> <nil> <nil> 0xc000668f00 <nil>}} 
>   < Exit [BeforeEach] test ifacer - /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:28 @ 10/26/23 14:37:48.96 (376ms)
>   > Enter [It] About ifacer's e2e use cases - /home/runner/work/spiderpool/spiderpool/test/e2e/ifacer/ifacer_test.go:101 @ 10/26/23 14:37:48.965
>   Create spidermultus config ns3748-584404833/test-multus-f3a984c7fd 
>   Try to create Deployment: ns3748-584404833/ds-5e08ac89f3 
>   Check that each node where the Pod is located should have a vlan sub-interface.
>   docker exec -i spiderpool1026142530-worker ip link show eth0 | grep 'state UP mode'  
>   docker exec -i spiderpool1026142530-worker ip link show | grep eth0.58 | grep 'state UP mode'  
>   Failed to execute command on the node, error is  
>   docker exec -i spiderpool1026142530-worker ip link show eth0 | grep 'state UP mode'  
>   docker exec -i spiderpool1026142530-worker ip link show | grep eth0.58 | grep 'state UP mode'  
>   docker exec -i spiderpool1026142530-control-plane ip link show eth0 | grep 'state UP mode'  
>   docker exec -i spiderpool1026142530-control-plane ip link show | grep eth0.58 | grep 'state UP mode'  
>   Create a vlan sub-interface with the same name, its network card status is down, and it is automatically set to up
>   docker exec -i spiderpool1026142530-worker ip link set eth0.58 down 
>   docker exec -i spiderpool1026142530-control-plane ip link set eth0.58 down 
>   [FAILED] Failed to execute down nic command on the node, error is Cannot find device "eth0.58"
> ```

Same as #2447, occasional problem, will be fixed as soon as possible

**Pull request merged**